core/threaded/watchdog.c

```diff
@@ -2,83 +2,173 @@
- * sleeping, so the watchdog never expires and the handler function is never
- * invoked.
- * This function acquires the reaction mutex and releases it while sleeping.
- * 
+ * sleeping, so when the thread wakes up, it can go back to sleep again. If the
+ * watchdog does expire. It will execute the watchdog handler and wait for next
+ * start. To stop a running watchdog, the expiration field is set to NEVER.
+ * The watchdog thread can also be terminated, this should only be done at the
+ * end of execution as it cannot be restarted after termination.
+ *
  * @param arg A pointer to the watchdog struct
  * @return NULL
  */
-void* _lf_run_watchdog(void* arg) {
+static void* watchdog_thread_main(void* arg) {
     watchdog_t* watchdog = (watchdog_t*)arg;
-
     self_base_t* base = watchdog->base;
-    assert(base->reactor_mutex != NULL);
+    LF_PRINT_DEBUG("Starting Watchdog %p", (void *) watchdog);
+    LF_ASSERT(base->reactor_mutex, "reactor-mutex not alloc'ed but has watchdogs.");
+
+    // Grab reactor-mutex and start infinite loop.
     LF_MUTEX_LOCK((lf_mutex_t*)(base->reactor_mutex));
-    instant_t physical_time = lf_time_physical();
-    while (physical_time < watchdog->expiration) {
-        interval_t T = watchdog->expiration - physical_time;
-        LF_MUTEX_UNLOCK((lf_mutex_t*)base->reactor_mutex);
-        lf_sleep(T);
-        LF_MUTEX_LOCK((lf_mutex_t*)(base->reactor_mutex));
-        physical_time = lf_time_physical();
-    }
+    while (true) {
 
-    if (watchdog->expiration != NEVER) {
+        // Step 1: Wait for a timeout to start watching for.
+        
+        // Edge case 1: We have already gotten a signal to terminate.
+        if (watchdog->terminate) {
+            goto terminate;
+        }
+
+        // Edge case 2: We have already received a timeout.
+        if(watchdog->expiration != NEVER) {
+            watchdog_wait(watchdog);
+        } else {
+            // Wait for a signal that we have a timeout to wait for on the cond-var.
+            LF_COND_WAIT(&watchdog->cond);
+            
+            // Check whether we actually got a termination signal.
+            if (watchdog->terminate) {
+                goto terminate;
+            }
+
+            // Finally go wait for that timeout.
+            watchdog_wait(watchdog);
+        }
+
+        // At this point we have returned from the watchdog wait. But it could
+        // be that it was to terminate the watchdog.
+        if (watchdog->terminate) {
+            goto terminate;
+        }
+
+        // It could also be that the watchdog was just stopped
+        if (watchdog->expiration == NEVER) {
+            continue;
+        }
+
+        // Finally. The watchdog actually timed out. Handle it.
+        LF_PRINT_DEBUG("Watchdog %p timed out", (void *) watchdog);
         watchdog_function_t watchdog_func = watchdog->watchdog_function;
         (*watchdog_func)(base);
+        watchdog->expiration = NEVER;
+        watchdog->active = false;
```

Comment:
```suggestion
        if(watchdog->expiration == NEVER) {
            // Watchdog has been stopped.
            // Wait for a signal that we have a timeout to wait for on the cond-var.
            LF_COND_WAIT(&watchdog->cond);
            continue;
        } else {
            // Watchdog has been started.
            watchdog_wait(watchdog);
            
            // At this point we have returned from the watchdog wait. But it could
            // be that it was to terminate the watchdog.
            if (watchdog->terminate) break;

               // It could also be that the watchdog was just stopped
            if (watchdog->expiration == NEVER) continue;

            // If we reach here, the watchdog actually timed out. Handle it.
            LF_PRINT_DEBUG("Watchdog %p timed out", (void *) watchdog);
            watchdog_function_t watchdog_func = watchdog->watchdog_function;
                (*watchdog_func)(base);
            watchdog->expiration = NEVER;
            watchdog->active = false;
        }
```